File: power-square-android/app/src/main/java/com/okla/ops/views/map/GoogleGeolocation.java
package com.okla.ops.views.map;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Location;
import android.os.Bundle;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;

import com.base.common.beans.LocationData;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;

import java.util.List;

public class GoogleGeolocation implements IGeolocation, GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {
    private Context mContext;
    private GoogleApiClient mGoogleApiClient;
    private static final LocationRequest REQUEST = LocationRequest.create()
            .setInterval(5000) // 5 seconds
            .setFastestInterval(4000) // 16ms = 60fps
            .setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
    public FusedLocationProviderClient mFusedLocationProviderClient;
    private LocationCallback mListener;

    com.google.android.gms.location.LocationCallback mLocationCallback = new com.google.android.gms.location.LocationCallback() {
        @Override
        public void onLocationResult(LocationResult locationResult) {
            List<Location> locationList = locationResult.getLocations();
//            Log.i(TAG,"onLocationResult success ...");
            if (!locationList.isEmpty()) {
                //The last location in the list is the newest
                Location googleLocation = locationList.get(locationList.size() - 1);

                //按存放顺序取数据
                LocationData data = new LocationData();
                data.setLatitude(googleLocation.getLatitude());//纬度
                data.setLongitude(googleLocation.getLongitude());//经度
                if (mListener != null)
                    mListener.onLocationChanged(data);
            }
        }
    };

    public GoogleGeolocation(Context context) {
        super();
        this.mContext = context;
    }

    @Override
    public void initLocationClient() {
        mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(mContext);
        mGoogleApiClient = new GoogleApiClient.Builder(mContext)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this)
                .addApi(LocationServices.API)
                .build();
        mGoogleApiClient.connect();
    }

    @Override
    public void onStartLocate(LocationCallback listener) {
        mListener = listener;

    }

    @Override
    public void onPauseLocate() {
        if (mFusedLocationProviderClient != null) {
            mFusedLocationProviderClient.removeLocationUpdates(mLocationCallback);
        }
    }

    @Override
    public void onStopLocate() {
        if (mFusedLocationProviderClient != null) {
            mFusedLocationProviderClient.removeLocationUpdates(mLocationCallback);
        }
    }

    @Override
    public void onConnected(Bundle bundle) {
        if (ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION)
                != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        mFusedLocationProviderClient.requestLocationUpdates(REQUEST, mLocationCallback, Looper.myLooper());
    }

    @Override
    public void onConnectionSuspended(int i) {
        if (mGoogleApiClient != null) {
            mGoogleApiClient.connect();
        }
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
    }

}

File: power-square-android/common/src/main/java/com/base/common/map/location/GoogleGeolocation.kt
package com.base.common.map.location

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.os.Looper
import androidx.core.app.ActivityCompat
import com.base.common.BuildConfig
import com.google.android.gms.location.*
import com.base.common.beans.LocationData
import com.base.common.map.IBaseGeoLocation
import com.base.common.map.GeoLocationCallback

class GoogleGeolocation(context: Context) : IBaseGeoLocation {
    val mContext = context
    lateinit var mListener: GeoLocationCallback
    lateinit var mFusedLocationProviderClient: FusedLocationProviderClient
    lateinit var mLocationCallback: LocationCallback
    lateinit var locationRequest: LocationRequest
    override fun initLocationClient() {
        locationRequest = LocationRequest.create()
            .setInterval(5000)
            .setFastestInterval(4000)
            .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
        mLocationCallback = object : LocationCallback() {
            override fun onLocationResult(result: LocationResult) {
                super.onLocationResult(result)
                val locationList: List<Location> = result.locations
                if (locationList.isNotEmpty()) {
                    val location: Location = locationList.last()

                    mListener.onLocationChanged(
                        LocationData(
                            location.latitude,
                            location.longitude
                        )
                    )
                }

            }
        }
        mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(mContext)
    }

    override fun onStartLocate(listener: GeoLocationCallback, minTimeMs: Long, minDistanceM: Float) {
        mListener = listener
        if (ActivityCompat.checkSelfPermission(
                mContext,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED ||
            ActivityCompat.checkSelfPermission(
                mContext,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            Looper.myLooper()?.let {
                mFusedLocationProviderClient.requestLocationUpdates(
                    locationRequest,
                    mLocationCallback,
                    it
                )
                    .addOnSuccessListener {
                        if (BuildConfig.DEBUG) {
                            mListener.onLocationChanged(
                                LocationData(
                                    22.580991666666666,
                                    113.90715
                                )
                            )
                        }
                    }.addOnFailureListener {
                    }.addOnCanceledListener {
                    }
            }
        }
    }

    override fun onPauseLocate() {
        mFusedLocationProviderClient.removeLocationUpdates(mLocationCallback)
    }

    override fun onStopLocate() {
        mFusedLocationProviderClient.removeLocationUpdates(mLocationCallback)
    }
}